### PR TITLE
Stage / Phase RC Bump 

### PR DESCRIPTION
**Pull request by davidbiancolin**
This will be squashed before preforming the FireSim submodule chicanery. This is ready for review and should be merged ASAP due to the number of submodules it touches.

Associated PRs:
- riscv-boom/riscv-boom#453
- firesim/firesim#532
- ucb-bar/testchipip#76
- ucb-bar/barstools#79
- ucb-bar/ariane-wrapper#3
- ucb-bar/hwacha#18
- firesim/icenet#23

Discussion below:

Here are the issues i'm running into that require some thought to port.

Challenges: 
1) Reset inference and IOCells.
  - DebugIO has two abstract resets. We need to decide how we wish to concretize these in our generator, before (or as part of) IOcell generation, as we cannot emit IOCell BBs with abstract resets for outputs.
  - Workaround: temporarily skip emission of IOCells for DebugIO 
  - Proposed solution: IOCell generation should croak on abstract resets. Add a case for AsyncReset, which uses the same underlying verilog but placates reset inference.

2) Test-suite generation:
  - Test-suite generation phases as they currently exist within rocket chip are not extensible 
  - Workaround: copy relevant rocket chip stage sources into chipyard and manually add our own phase for this purpose
  - Proposed solution A: Let the generator emit annotations calling for specific test suites instead of having a phase to do so. Collect these in a later phase. 
-  Proposed solution B: Make it possible to add additional test suites in the existing phase / make it possible to run our own phase which does this.

3) Generator command-line API
  - The new rocket-chip phase removes the "_" delimited configs in favor of comma-delimited ones
  - Workaround: We're copying stage sources into chipyard anyways, so in our stage let's just include the legacy mode.
- Proposed solution: -Upstream a --legacy-config -LC to rocketchip 

TODOs:
- [x] Rectify make variable hacks to maintain existing make API 
- [x] Fix longname

### CODE REVIEW

**Comment by jwright6323** [comment]
- [ ] Just noting here that when we go to switch `IOBinders` to matching on the `LazyModule` instead of the `Module`, we should put the `system` `suggestName` inside `ChipTop` (#480)

File: generators/chipyard/src/main/scala/Subsystem.scala [hunk 1]
```diff
@@ -52,25 +52,37 @@ trait HasChipyardTiles extends HasTiles
...
+      //    val t = LazyModule(new ArianeTile(a, crossing, PriorityMuxHartIdFromSeq(arianeTileParams), logicalTreeNode))
+      //    (t, t.rocketLogicalTree) // TODO FIX rocketLogicalTree is not a member of the superclass, both child classes define it separately
+      //  }
+      //}
       connectMasterPortsToSBus(tile, crossing)
       connectSlavePortsToCBus(tile, crossing)
 
-      def treeNode: RocketTileLogicalTreeNode = new RocketTileLogicalTreeNode(rocketLogicalTree.getOMInterruptTargets)
-      LogicalModuleTree.add(logicalTreeNode, rocketLogicalTree)
+      //def treeNode: RocketTileLogicalTreeNode = new RocketTileLogicalTreeNode(rocketLogicalTree.getOMInterruptTargets)
```
Comment (abejgonzalez) [inline generators/chipyard/src/main/scala/Subsystem.scala:84]: Why is this deleted?

Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/Subsystem.scala:84]: We don't use the this OM so i think we should remove our support for it.

**Comment by abejgonzalez** [comment]
@zhemao Why does IceNet need a `TestHarness` file? Is it still needed?

File: generators/chipyard/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (abejgonzalez) [reply generators/chipyard/src/main/scala/Subsystem.scala:84]: My impression when I last looked at it a while ago was that it was tied to the DTS. I'm not opposed to getting rid of support for it as long as it doesn't affect anything else.

File: variables.mk [hunk 2]
```diff
@@ -86,7 +86,7 @@ CHIPYARD_FIRRTL_DIR = $(base_dir)/tools/firrtl
 #########################################################################################
 # names of various files needed to compile and run things
 #########################################################################################
-long_name = $(MODEL_PACKAGE).$(MODEL).$(CONFIG)
+long_name = $(CONFIG_PACKAGE).$(MODEL).$(CONFIG)
```
Comment (abejgonzalez) [inline variables.mk:89]: Random thought: Should we look to re-organize this into something like "MODEL.CONFIG_PACKAGE.CONFIG"? Just to have it be clear that the CONFIG comes from CONFIG_PACKAGE and the MODEL can be located somewhere else?

Comment (abejgonzalez) [reply variables.mk:89]: Also, should we start deprecating the support for the Hwacha generator @colinschmidt? Is there a use case for it to still exist?

File: generators/chipyard/src/main/scala/System.scala [hunk 3]
```diff
@@ -38,8 +38,5 @@ class System(implicit p: Parameters) extends Subsystem
 class SystemModule[+L <: System](_outer: L) extends SubsystemModuleImp(_outer)
   with HasRTCModuleImp
   with HasExtInterruptsModuleImp
-  with CanHaveMasterAXI4MemPortModuleImp
```
Comment (abejgonzalez) [inline generators/chipyard/src/main/scala/System.scala:41]: Why is this deleted here? Am I missing something / where they are added?

File: variables.mk (hunk 2 above)
Comment (davidbiancolin) [reply variables.mk:89]: I think that makes more sense.

File: generators/chipyard/src/main/scala/System.scala (hunk 3 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/System.scala:41]: There's no imp trait anymore; it's been replaced with an InModuleBody call

File: generators/chipyard/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/Subsystem.scala:84]: ill check that the generated DTS is the same

**Comment by colinschmidt** [comment]
For the test suite generation I like the idea of emitting annotations, but not exactly sure how it would work out.
I also like `--legacy-config`

File: build.sbt [hunk 4]
```diff
@@ -166,6 +166,8 @@ lazy val gemmini = (project in file("generators/gemmini"))
 lazy val tapeout = conditionalDependsOn(project in file("./tools/barstools/tapeout/"))
   .dependsOn(chisel_testers, chipyard)
   .settings(commonSettings)
+  // Biancolin Bump: How did things compile without this dependency previously?
+  .settings(libraryDependencies ++= Seq("io.github.daviddenton" %% "handlebars-scala-fork" % "2.3.0"))
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply build.sbt:170]: @grebe might know

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 5]
```diff
@@ -12,7 +12,7 @@ import chipyard.iobinders.{IOBinders, TestHarnessFunction, IOBinderTuple}
 
 import barstools.iocell.chisel._
 
-case object BuildSystem extends Field[Parameters => RawModule]((p: Parameters) => Module(LazyModule(new DigitalTop()(p)).suggestName("system").module))
+case object BuildSystem extends Field[Parameters => LazyModule]((p: Parameters) => LazyModule(new DigitalTop()(p)).suggestName("system"))
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ChipTop.scala:15]: can we move the `suggestName` from here to `ChipTop` now?

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 6]
```diff
@@ -31,11 +31,15 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
   val systemReset = Wire(Input(Bool()))
 
   // The system module specified by BuildSystem
-  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+  val lSystem = p(BuildSystem)(p)
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ChipTop.scala:34]:
```suggestion
  val lSystem = p(BuildSystem)(p).suggestName("system")
```

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 5 above)
Comment (jwright6323) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: Also, is there ever a case where someone would want to have a concrete `Module` instead of a `LazyModule` as the system? It's easy to go from `LazyModule` -> `RawModule` but the reverse  requires a bit of work. Maybe that's rare enough that it's fine.

File: variables.mk (hunk 2 above)
Comment (jerryz123) [reply variables.mk:89]: I'm in favor of removing `SUB_PROJECT`.

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 7]
```diff
@@ -31,11 +31,15 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
...
+  val system = withClockAndReset(systemClock, systemReset) { Module(lSystem.module) }
 
   // Call all of the IOBinders and provide them with a default clock and reset
   withClockAndReset(systemClock, systemReset) {
-    val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.map(_(system)).flatten.unzip3
+    // Biancolin: I'm lazy and didn't want to port code, but this may be a feature...
```
[the remark this answers is not in the document]
Comment (jerryz123) [reply generators/chipyard/src/main/scala/ChipTop.scala:39]: :scream: despite how this makes me uncomfortable, the flexibility seems to be useful enough.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 5 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: Why is the reverse hard? It should be pretty easy to create a empty `LazyModule` to wrap an arbitrary concrete `Module`

Comment (jwright6323) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: Eh, "Hard" is probably the wrong word. My point was that creating an entirely new class is more typing than 8 characters (`.module`).

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:39]: I'm happy to add a check here which only calls it on the LazyModule if the PF is defined on both.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 5 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: We could have both a buildTop and buildLazy top if we really wanted and delegate appropriately, or we could have the BuildTop function return `(Option[LazyModule], Module)`

Comment (jwright6323) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: `(Option[LazyModule], Module)` actually seems pretty good.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (colinschmidt) [reply generators/chipyard/src/main/scala/ChipTop.scala:39]: Yeah I like that check. 👍

File: generators/chipyard/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/Subsystem.scala:84]: @abejgonzalez The DTS seems unaffected by this change

File: build.sbt (hunk 4 above)
[the remark this answers is not in the document]
Comment (grebe) [reply build.sbt:170]: `conditionalDependsOn` discards the build.sbt info b/c sbt is dumb at handling subprojects (I think the problematic case is actually subprojects of subprojects. I forget. Rocketchip in my experience has been very difficult to get working without entirely dodging the problem). Chipyard's top level build.sbt needs to reproduce all the deps from subprojects that use `conditionalDependsOn`. This looks like something barstools needs

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:39]: Ok so it's probably going to be like 10 - 15 lines of reflection to make this happen ... so i'll  write it in a later PR.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 5 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]: I think this might require more careful thought so i'll resolve it in a follow up PR.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 6 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:34]: I've applied this to a local change i'll push tomorrow.

File: generators/chipyard/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (hcook) [reply generators/chipyard/src/main/scala/Subsystem.scala]: They're (unfortunately) parallel implementations that share some common infrastructure. If anything, OM builds/relies on the `Device` snippets used by DTS, not vice versa.

File: variables.mk (hunk 2 above)
Comment (abejgonzalez) [reply variables.mk]: Likewise at the beginning it made sense. Now less so.

File: generators/chipyard/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (abejgonzalez) [reply generators/chipyard/src/main/scala/Subsystem.scala]: Thanks for the clarification

File: variables.mk (hunk 2 above)
Comment (abejgonzalez) [reply variables.mk]: @colinschmidt Thoughts on the Hwacha generator? If you don't use it then we can just use the Chipyard generator and get rid of the "SUB_PROJECT=hwacha"

Comment (colinschmidt) [reply variables.mk]: It needs to provide feature parity.

Comment (davidbiancolin) [reply variables.mk]: Unless i screwed up porting the Makefrag generation, it should now.

Comment (colinschmidt) [reply variables.mk]: Looks reasonable to me. I'll test after the merge and fix it if its broken.

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 5 above)
Comment (jwright6323) [reply generators/chipyard/src/main/scala/ChipTop.scala:15]:
```suggestion
case object BuildSystem extends Field[Parameters => LazyModule]((p: Parameters) => LazyModule(new DigitalTop()(p)))
```

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 8]
```diff
@@ -26,16 +61,21 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
...
+  // These are given so that IOCell can use DataMirror and generate ports with
+  // the right flow (Input/Output)
   val systemClock = Wire(Input(Clock()))
-  // The system reset (synchronous to clock)
-  val systemReset = Wire(Input(Bool()))
+  val systemReset = Wire(Input(Reset()))
 
   // The system module specified by BuildSystem
-  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+  val lSystem = p(BuildSystem)(p).suggestName("system")
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply generators/chipyard/src/main/scala/ChipTop.scala:70]: @zhemao I think we wanted to get rid of the other one

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 9]
```diff
@@ -45,13 +85,22 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
 
 /**
  * A simple clock and reset implementation that punches out clock and reset ports with the same
- * names as the implicit clock and reset for standard Module classes. Reset is synchronous to
- * clock, which may not be a good idea to use for tapeouts.
+ * names as the implicit clock and reset for standard Module classes. Three basic reset schemes 
+ * a provided. See [[GlobalResetScheme]].
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ChipTop.scala:89]:
```suggestion
 * are provided. See [[GlobalResetScheme]].
```

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 10]
```diff
@@ -45,13 +85,22 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
...
  */
 trait HasChipTopSimpleClockAndReset { this: BaseChipTop =>
 
   val (clock, systemClockIO) = IOCell.generateIOFromSignal(systemClock, Some("iocell_clock"))
-  val (reset, systemResetIO) = IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"))
+  val (reset, systemResetIO) = p(GlobalResetSchemeKey) match {
+    case GlobalResetSynchronous  =>
+      IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"))
+    case GlobalResetAsynchronousFull =>
+      IOCell.generateIOFromSignal(systemReset, Some("iocell_areset"), abstractResetAsAsync = true)
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ChipTop.scala:98]:
```suggestion
      IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"), abstractResetAsAsync = true)
```

Unless we're going to suggestName `reset` to `areset` we probably want to keep the IO cell naming convention of `iocell_ + <signal name>`

File: generators/chipyard/src/main/scala/ChipTop.scala [hunk 11]
```diff
@@ -45,13 +85,22 @@ abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with
...
+    case GlobalResetAsynchronous =>
+      val asyncResetCore = Wire(Input(AsyncReset()))
+      systemReset := ResetCatchAndSync(systemClock, asyncResetCore.asBool)
+      IOCell.generateIOFromSignal(asyncResetCore, Some("iocell_areset"), abstractResetAsAsync = true)
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ChipTop.scala:102]:
```suggestion
      IOCell.generateIOFromSignal(asyncResetCore, Some("iocell_reset"), abstractResetAsAsync = true)
```

See above

File: generators/chipyard/src/main/scala/ConfigFragments.scala [hunk 12]
```diff
@@ -66,7 +66,7 @@ class WithL2TLBs(entries: Int) extends Config((site, here, up) => {
 })
 
 class WithTracegenSystem extends Config((site, here, up) => {
-  case BuildSystem => (p: Parameters) => Module(LazyModule(new tracegen.TraceGenSystem()(p)).suggestName("Top").module)
+  case BuildSystem => (p: Parameters) => LazyModule(new tracegen.TraceGenSystem()(p)).suggestName("Top")
```
Comment (jwright6323) [inline generators/chipyard/src/main/scala/ConfigFragments.scala:69]:
```suggestion
  case BuildSystem => (p: Parameters) => LazyModule(new tracegen.TraceGenSystem()(p))
```

File: generators/chipyard/src/main/scala/ChipTop.scala (hunk 10 above)
Comment (davidbiancolin) [reply generators/chipyard/src/main/scala/ChipTop.scala:98]: sounds good.

**Comment by abejgonzalez** [comment]
LGTM pending Ariane/BOOM comments.
